libraries/tools/kotlin-main-kts/src/org/jetbrains/kotlin/mainKts/scriptDef.kt

```diff
@@ -90,6 +94,72 @@ class MainKtsHostConfiguration : ScriptingHostConfiguration(
     }
 )
 
+fun configureScriptFileLocationPathVariablesForEvaluation(context: ScriptEvaluationConfigurationRefinementContext): ResultWithDiagnostics<ScriptEvaluationConfiguration> {
+    val scriptId = context.compiledScript.sourceLocationId ?: return context.evaluationConfiguration.asSuccess()
+    val compilationConfiguration = context.evaluationConfiguration[ScriptEvaluationConfiguration.compilationConfiguration]
+        ?: return context.evaluationConfiguration.asSuccess()
+    val scriptFileLocationMap = compilationConfiguration[ScriptCompilationConfiguration.scriptFileLocation]
+        ?: return context.evaluationConfiguration.asSuccess()
+    val scriptFileLocation = scriptFileLocationMap[scriptId] ?: return context.evaluationConfiguration.asSuccess()
+
+    val scriptFileLocationVariable =
+        compilationConfiguration[ScriptCompilationConfiguration.scriptFileLocationVariables]?.get(scriptId)
+
+    val res = if (scriptFileLocationVariable != null) {
+        context.evaluationConfiguration.with {
+            providedProperties.put(mapOf(scriptFileLocationVariable to scriptFileLocation))
+        }
+    } else context.evaluationConfiguration
+    return res.asSuccess()
+}
+
+fun configureScriptFileLocationPathVariablesForCompilation(context: ScriptConfigurationRefinementContext): ResultWithDiagnostics<ScriptCompilationConfiguration> {
+    val scriptFile = (context.script as? FileBasedScriptSource)?.file ?: return context.compilationConfiguration.asSuccess()
```

Comment:
Possible problem with the approach of skipping the variable if it is not possible to determine the location: it will not be possible to write a generic script that could check if it is loaded from a file or not. 
It would be better to still define `__FILE__` to `null` or empty/root file in this case. (I guess it will be a bit problematic to add a nullable variable, seems like it's another blind spot in the API.) 
Probably there is no need to fix it in this PR, and just something to keep in mind for the future improvements.

Reply:
I agree with the point about `null`. Having an empty/root file looks a bit less clean to me to be honest. It would indeed be ideal if in the future nullable variables are supported :)